File: isp_management-main/docs/api/client_examples/isp_api_client.py
"""
ISP Management Platform API Client

A Python client library for interacting with the ISP Management Platform API.
This client takes advantage of HATEOAS links, handles rate limiting, and provides
standardized error handling.
"""

import time
import json
import urllib.parse
from typing import Dict, List, Any, Optional, Union
import requests


class ApiError(Exception):
    """Custom exception for API errors with detailed information."""
    
    def __init__(self, message: str, status: int, code: int = None, details: Dict = None):
        self.message = message
        self.status = status
        self.code = code or status
        self.details = details or {}
        super().__init__(self.message)
    
    def get_field_error(self, field_name: str) -> Optional[str]:
        """Get error message for a specific field."""
        if not self.details or 'field_errors' not in self.details:
            return None
        
        for field_error in self.details['field_errors']:
            if field_error['field'] == field_name:
                return field_error['message']
        
        return None


class IspApiClient:
    """
    Client for the ISP Management Platform API.
    
    This client handles authentication, rate limiting, and provides methods
    for interacting with the API resources. It also takes advantage of HATEOAS
    links for navigation.
    """
    
    def __init__(self, base_url: str, api_key: str):
        """
        Initialize the API client.
        
        Args:
            base_url: The base URL of the API (e.g., 'https://api.example.com')
            api_key: API key for authentication
        """
        self.base_url = base_url.rstrip('/')
        self.api_key = api_key
        self.rate_limits = {
            'limit': float('inf'),
            'remaining': float('inf'),
            'reset': 0
        }
        
        # Cache for discovered links
        self.link_cache = {}
        
        # Session for connection pooling
        self.session = requests.Session()
        self.session.headers.update({
            'Authorization': f'Bearer {self.api_key}',
            'Content-Type': 'application/json',
            'Accept': 'application/json'
        })
    
    def request(self, url: str, method: str = 'GET', data: Dict = None, 
                params: Dict = None, headers: Dict = None) -> Dict:
        """
        Make an API request with automatic rate limit handling.
        
        Args:
            url: The URL to request
            method: HTTP method (GET, POST, PUT, DELETE)
            data: Request body data
            params: Query parameters
            headers: Additional headers
            
        Returns:
            The response data
            
        Raises:
            ApiError: If the API returns an error response
        """
        # Check if we're rate limited
        if self.rate_limits['remaining'] <= 0:
            now = int(time.time())
            wait_time = max(0, self.rate_limits['reset'] - now)
            
            if wait_time > 0:
                print(f"Rate limit reached. Waiting {wait_time} seconds before retrying.")
                time.sleep(wait_time)
        
        # Prepare the request
        full_url = url if url.startswith('http') else f"{self.base_url}{url}"
        request_headers = headers or {}
        
        # Make the request
        response = self.session.request(
            method=method,
            url=full_url,
            json=data if data else None,
            params=params,
            headers=request_headers
        )
        
        # Update rate limit information
        self._update_rate_limits(response)
        
        # Parse the response
        try:
            response_data = response.json()
        except ValueError:
            response_data = {'message': 'Invalid JSON response'}
        
        # Handle error responses
        if response.status_code >= 400:
            raise self._create_error_from_response(response_data, response.status_code)
        
        # Cache any links in the response
        if '_links' in response_data:
            self._cache_links(url, response_data['_links'])
        
        return response_data
    
    def _update_rate_limits(self, response: requests.Response) -> None:
        """
        Update rate limit information from response headers.
        
        Args:
            response: Requests Response object
        """
        limit = response.headers.get('X-RateLimit-Limit')
        remaining = response.headers.get('X-RateLimit-Remaining')
        reset = response.headers.get('X-RateLimit-Reset')
        
        if limit:
            self.rate_limits['limit'] = int(limit)
        if remaining:
            self.rate_limits['remaining'] = int(remaining)
        if reset:
            self.rate_limits['reset'] = int(reset)
    
    def _create_error_from_response(self, data: Dict, status: int) -> ApiError:
        """
        Create a standardized error object from an API error response.
        
        Args:
            data: Error response data
            status: HTTP status code
            
        Returns:
            Enhanced ApiError object
        """
        return ApiError(
            message=data.get('message', 'API request failed'),
            status=status,
            code=data.get('code', status),
            details=data.get('details', {})
        )
    
    def _cache_links(self, url: str, links: Dict) -> None:
        """
        Cache links from a response for later use.
        
        Args:
            url: The URL that was requested
            links: The _links object from the response
        """
        self.link_cache[url] = links
    
    def follow_link(self, resource: Dict, rel: str, method: str = None, 
                   data: Dict = None, params: Dict = None) -> Dict:
        """
        Follow a link from a previous response.
        
        Args:
            resource: Resource containing _links
            rel: The link relation to follow
            method: HTTP method override
            data: Request body data
            params: Query parameters
            
        Returns:
            The response data
            
        Raises:
            ValueError: If the link relation is not found
        """
        if '_links' not in resource or rel not in resource['_links']:
            raise ValueError(f"Link relation '{rel}' not found in resource")
        
        link = resource['_links'][rel]
        return self.request(
            url=link['href'],
            method=method or link.get('method', 'GET'),
            data=data,
            params=params
        )
    
    # Customer endpoints
    
    def get_customers(self, **params) -> Dict:
        """
        Get a list of customers with optional filtering.
        
        Args:
            **params: Query parameters (skip, limit, customer_type, etc.)
            
        Returns:
            Paginated customer list with HATEOAS links
        """
        url = '/api/v1/customers'
        return self.request(url, params=params)
    
    def get_customer(self, customer_id: Union[str, int]) -> Dict:
        """
        Get a customer by ID.
        
        Args:
            customer_id: Customer ID
            
        Returns:
            Customer data with HATEOAS links
        """
        url = f'/api/v1/customers/{customer_id}'
        return self.request(url)
    
    def create_customer(self, customer_data: Dict) -> Dict:
        """
        Create a new customer.
        
        Args:
            customer_data: Customer data
            
        Returns:
            Created customer with HATEOAS links
        """
        url = '/api/v1/customers'
        return self.request(url, method='POST', data=customer_data)
    
    def update_customer(self, customer_id: Union[str, int], customer_data: Dict) -> Dict:
        """
        Update a customer.
        
        Args:
            customer_id: Customer ID
            customer_data: Updated customer data
            
        Returns:
            Updated customer with HATEOAS links
        """
        url = f'/api/v1/customers/{customer_id}'
        return self.request(url, method='PUT', data=customer_data)
    
    def delete_customer(self, customer_id: Union[str, int]) -> Dict:
        """
        Delete a customer.
        
        Args:
            customer_id: Customer ID
            
        Returns:
            Response data
        """
        url = f'/api/v1/customers/{customer_id}'
        return self.request(url, method='DELETE')
    
    # Customer related resources
    
    def get_customer_addresses(self, customer: Dict) -> Dict:
        """
        Get customer addresses using HATEOAS link.
        
        Args:
            customer: Customer resource with _links
            
        Returns:
            Customer addresses with HATEOAS links
        """
        return self.follow_link(customer, 'addresses')
    
    def get_customer_contacts(self, customer: Dict) -> Dict:
        """
        Get customer contacts using HATEOAS link.
        
        Args:
            customer: Customer resource with _links
            
        Returns:
            Customer contacts with HATEOAS links
        """
        return self.follow_link(customer, 'contacts')
    
    def get_customer_documents(self, customer: Dict) -> Dict:
        """
        Get customer documents using HATEOAS link.
        
        Args:
            customer: Customer resource with _links
            
        Returns:
            Customer documents with HATEOAS links
        """
        return self.follow_link(customer, 'documents')
    
    def get_customer_subscription(self, customer: Dict) -> Dict:
        """
        Get customer subscription using HATEOAS link.
        
        Args:
            customer: Customer resource with _links
            
        Returns:
            Customer subscription with HATEOAS links
        """
        return self.follow_link(customer, 'subscription')
    
    def get_customer_billing(self, customer: Dict) -> Dict:
        """
        Get customer billing information using HATEOAS link.
        
        Args:
            customer: Customer resource with _links
            
        Returns:
            Customer billing with HATEOAS links
        """
        return self.follow_link(customer, 'billing')
    
    # Pagination helpers
    
    def get_next_page(self, collection: Dict) -> Dict:
        """
        Navigate to the next page of a collection.
        
        Args:
            collection: Collection resource with _links
            
        Returns:
            Next page of the collection
        """
        return self.follow_link(collection, 'next')
    
    def get_previous_page(self, collection: Dict) -> Dict:
        """
        Navigate to the previous page of a collection.
        
        Args:
            collection: Collection resource with _links
            
        Returns:
            Previous page of the collection
        """
        return self.follow_link(collection, 'prev')
    
    def get_first_page(self, collection: Dict) -> Dict:
        """
        Navigate to the first page of a collection.
        
        Args:
            collection: Collection resource with _links
            
        Returns:
            First page of the collection
        """
        return self.follow_link(collection, 'first')
    
    def get_last_page(self, collection: Dict) -> Dict:
        """
        Navigate to the last page of a collection.
        
        Args:
            collection: Collection resource with _links
            
        Returns:
            Last page of the collection
        """
        return self.follow_link(collection, 'last')
    
    # Monitoring endpoints
    
    def get_alert_configurations(self, **params) -> Dict:
        """
        Get alert configurations with optional filtering.
        
        Args:
            **params: Query parameters (skip, limit, service_name, is_active)
            
        Returns:
            Paginated alert configurations with HATEOAS links
        """
        url = '/api/v1/monitoring/alert-configurations'
        return self.request(url, params=params)
    
    def get_alert_history(self, **params) -> Dict:
        """
        Get alert history with optional filtering.
        
        Args:
            **params: Query parameters (skip, limit, service_names, severities, etc.)
            
        Returns:
            Paginated alert history with HATEOAS links
        """
        url = '/api/v1/monitoring/alert-history'
        return self.request(url, params=params)


# Example usage
def example_usage():
    """Example of using the API client."""
    try:
        # Initialize the client
        client = IspApiClient('https://api.example.com', 'your-api-key')
        
        # Get customers (first page)
        customers = client.get_customers(limit=10)
        print(f"Found {customers['total']} customers")
        
        # Navigate to the next page using HATEOAS links
        if 'next' in customers.get('_links', {}):
            next_page = client.get_next_page(customers)
            print(f"Showing customers {next_page['skip'] + 1} to {next_page['skip'] + len(next_page['items'])}")
        
        # Get a specific customer
        customer = client.get_customer(123)
        print(f"Customer: {customer['name']}")
        
        # Get customer addresses using HATEOAS links
        addresses = client.get_customer_addresses(customer)
        print(f"Customer has {len(addresses['items'])} addresses")
        
        # Create a new customer
        new_customer = client.create_customer({
            'name': 'New Customer',
            'email': 'new@example.com'
        })
        print(f"Created customer with ID: {new_customer['id']}")
        
    except ApiError as e:
        print(f"API Error ({e.status}): {e.message}")
        
        # Handle validation errors
        if e.details and 'field_errors' in e.details:
            for field_error in e.details['field_errors']:
                print(f"{field_error['field']}: {field_error['message']}")


if __name__ == '__main__':
    example_usage()

File: isp_management-main/docs/api/client_examples/isp_api_client.js
/**
 * ISP Management Platform API Client
 * 
 * A JavaScript client library for interacting with the ISP Management Platform API.
 * This client takes advantage of HATEOAS links, handles rate limiting, and provides
 * standardized error handling.
 */

class IspApiClient {
  /**
   * Create a new API client instance
   * 
   * @param {string} baseUrl - The base URL of the API (e.g., 'https://api.example.com')
   * @param {string} apiKey - API key for authentication
   */
  constructor(baseUrl, apiKey) {
    this.baseUrl = baseUrl;
    this.apiKey = apiKey;
    this.rateLimits = {
      limit: Infinity,
      remaining: Infinity,
      reset: 0
    };
    
    // Cache for discovered links
    this.linkCache = {};
  }

  /**
   * Make an API request with automatic rate limit handling
   * 
   * @param {string} url - The URL to request
   * @param {Object} options - Fetch options
   * @returns {Promise<Object>} - The response data
   */
  async request(url, options = {}) {
    // Add authentication
    const headers = {
      'Authorization': `Bearer ${this.apiKey}`,
      'Content-Type': 'application/json',
      ...options.headers
    };

    // Check if we're rate limited
    if (this.rateLimits.remaining <= 0) {
      const now = Math.floor(Date.now() / 1000);
      const waitTime = Math.max(0, this.rateLimits.reset - now);
      
      if (waitTime > 0) {
        console.warn(`Rate limit reached. Waiting ${waitTime} seconds before retrying.`);
        await new Promise(resolve => setTimeout(resolve, waitTime * 1000));
      }
    }

    // Make the request
    const response = await fetch(url.startsWith('http') ? url : `${this.baseUrl}${url}`, {
      ...options,
      headers
    });

    // Update rate limit information
    this.updateRateLimits(response);

    // Parse the response
    const data = await response.json();

    // Handle error responses
    if (response.status >= 400) {
      throw this.createErrorFromResponse(data, response.status);
    }

    // Cache any links in the response
    if (data._links) {
      this.cacheLinks(url, data._links);
    }

    return data;
  }

  /**
   * Update rate limit information from response headers
   * 
   * @param {Response} response - Fetch Response object
   */
  updateRateLimits(response) {
    const limit = response.headers.get('X-RateLimit-Limit');
    const remaining = response.headers.get('X-RateLimit-Remaining');
    const reset = response.headers.get('X-RateLimit-Reset');

    if (limit) this.rateLimits.limit = parseInt(limit);
    if (remaining) this.rateLimits.remaining = parseInt(remaining);
    if (reset) this.rateLimits.reset = parseInt(reset);
  }

  /**
   * Create a standardized error object from an API error response
   * 
   * @param {Object} data - Error response data
   * @param {number} status - HTTP status code
   * @returns {Error} - Enhanced error object
   */
  createErrorFromResponse(data, status) {
    const error = new Error(data.message || 'API request failed');
    error.status = status;
    error.code = data.code || status;
    error.details = data.details || {};
    
    // Add helper for validation errors
    if (error.details.field_errors) {
      error.getFieldError = (fieldName) => {
        const fieldError = error.details.field_errors.find(e => e.field === fieldName);
        return fieldError ? fieldError.message : null;
      };
    }
    
    return error;
  }

  /**
   * Cache links from a response for later use
   * 
   * @param {string} url - The URL that was requested
   * @param {Object} links - The _links object from the response
   */
  cacheLinks(url, links) {
    this.linkCache[url] = links;
  }

  /**
   * Follow a link from a previous response
   * 
   * @param {Object} resource - Resource containing _links
   * @param {string} rel - The link relation to follow
   * @param {Object} options - Additional fetch options
   * @returns {Promise<Object>} - The response data
   */
  async followLink(resource, rel, options = {}) {
    if (!resource._links || !resource._links[rel]) {
      throw new Error(`Link relation '${rel}' not found in resource`);
    }

    const link = resource._links[rel];
    return this.request(link.href, {
      method: link.method || 'GET',
      ...options
    });
  }

  /**
   * Get a list of customers with optional filtering
   * 
   * @param {Object} params - Query parameters
   * @returns {Promise<Object>} - Paginated customer list with HATEOAS links
   */
  async getCustomers(params = {}) {
    const queryString = new URLSearchParams(params).toString();
    const url = `/api/v1/customers${queryString ? `?${queryString}` : ''}`;
    return this.request(url);
  }

  /**
   * Get a customer by ID
   * 
   * @param {string|number} id - Customer ID
   * @returns {Promise<Object>} - Customer data with HATEOAS links
   */
  async getCustomer(id) {
    return this.request(`/api/v1/customers/${id}`);
  }

  /**
   * Create a new customer
   * 
   * @param {Object} customerData - Customer data
   * @returns {Promise<Object>} - Created customer with HATEOAS links
   */
  async createCustomer(customerData) {
    return this.request('/api/v1/customers', {
      method: 'POST',
      body: JSON.stringify(customerData)
    });
  }

  /**
   * Update a customer
   * 
   * @param {string|number} id - Customer ID
   * @param {Object} customerData - Updated customer data
   * @returns {Promise<Object>} - Updated customer with HATEOAS links
   */
  async updateCustomer(id, customerData) {
    return this.request(`/api/v1/customers/${id}`, {
      method: 'PUT',
      body: JSON.stringify(customerData)
    });
  }

  /**
   * Delete a customer
   * 
   * @param {string|number} id - Customer ID
   * @returns {Promise<Object>} - Response data
   */
  async deleteCustomer(id) {
    return this.request(`/api/v1/customers/${id}`, {
      method: 'DELETE'
    });
  }

  /**
   * Get customer addresses
   * 
   * @param {Object} customer - Customer resource with _links
   * @returns {Promise<Object>} - Customer addresses with HATEOAS links
   */
  async getCustomerAddresses(customer) {
    return this.followLink(customer, 'addresses');
  }

  /**
   * Get customer contacts
   * 
   * @param {Object} customer - Customer resource with _links
   * @returns {Promise<Object>} - Customer contacts with HATEOAS links
   */
  async getCustomerContacts(customer) {
    return this.followLink(customer, 'contacts');
  }

  /**
   * Get customer documents
   * 
   * @param {Object} customer - Customer resource with _links
   * @returns {Promise<Object>} - Customer documents with HATEOAS links
   */
  async getCustomerDocuments(customer) {
    return this.followLink(customer, 'documents');
  }

  /**
   * Get customer subscription
   * 
   * @param {Object} customer - Customer resource with _links
   * @returns {Promise<Object>} - Customer subscription with HATEOAS links
   */
  async getCustomerSubscription(customer) {
    return this.followLink(customer, 'subscription');
  }

  /**
   * Get customer billing information
   * 
   * @param {Object} customer - Customer resource with _links
   * @returns {Promise<Object>} - Customer billing with HATEOAS links
   */
  async getCustomerBilling(customer) {
    return this.followLink(customer, 'billing');
  }

  /**
   * Navigate to the next page of a collection
   * 
   * @param {Object} collection - Collection resource with _links
   * @returns {Promise<Object>} - Next page of the collection
   */
  async getNextPage(collection) {
    return this.followLink(collection, 'next');
  }

  /**
   * Navigate to the previous page of a collection
   * 
   * @param {Object} collection - Collection resource with _links
   * @returns {Promise<Object>} - Previous page of the collection
   */
  async getPreviousPage(collection) {
    return this.followLink(collection, 'prev');
  }

  /**
   * Navigate to the first page of a collection
   * 
   * @param {Object} collection - Collection resource with _links
   * @returns {Promise<Object>} - First page of the collection
   */
  async getFirstPage(collection) {
    return this.followLink(collection, 'first');
  }

  /**
   * Navigate to the last page of a collection
   * 
   * @param {Object} collection - Collection resource with _links
   * @returns {Promise<Object>} - Last page of the collection
   */
  async getLastPage(collection) {
    return this.followLink(collection, 'last');
  }

  /**
   * Get alert configurations
   * 
   * @param {Object} params - Query parameters
   * @returns {Promise<Object>} - Paginated alert configurations with HATEOAS links
   */
  async getAlertConfigurations(params = {}) {
    const queryString = new URLSearchParams(params).toString();
    const url = `/api/v1/monitoring/alert-configurations${queryString ? `?${queryString}` : ''}`;
    return this.request(url);
  }

  /**
   * Get alert history
   * 
   * @param {Object} params - Query parameters
   * @returns {Promise<Object>} - Paginated alert history with HATEOAS links
   */
  async getAlertHistory(params = {}) {
    const queryString = new URLSearchParams(params).toString();
    const url = `/api/v1/monitoring/alert-history${queryString ? `?${queryString}` : ''}`;
    return this.request(url);
  }
}

// Example usage
async function exampleUsage() {
  try {
    // Initialize the client
    const client = new IspApiClient('https://api.example.com', 'your-api-key');
    
    // Get customers (first page)
    const customers = await client.getCustomers({ limit: 10 });
    console.log(`Found ${customers.total} customers`);
    
    // Navigate to the next page using HATEOAS links
    if (customers._links.next) {
      const nextPage = await client.getNextPage(customers);
      console.log(`Showing customers ${nextPage.skip + 1} to ${nextPage.skip + nextPage.items.length}`);
    }
    
    // Get a specific customer
    const customer = await client.getCustomer(123);
    console.log(`Customer: ${customer.name}`);
    
    // Get customer addresses using HATEOAS links
    const addresses = await client.getCustomerAddresses(customer);
    console.log(`Customer has ${addresses.items.length} addresses`);
    
    // Create a new customer
    const newCustomer = await client.createCustomer({
      name: 'New Customer',
      email: 'new@example.com'
    });
    console.log(`Created customer with ID: ${newCustomer.id}`);
    
  } catch (error) {
    console.error('API Error:', error.message);
    
    // Handle validation errors
    if (error.details && error.details.field_errors) {
      error.details.field_errors.forEach(fieldError => {
        console.error(`${fieldError.field}: ${fieldError.message}`);
      });
    }
  }
}

// Export the client class
if (typeof module !== 'undefined') {
  module.exports = { IspApiClient };
}
